File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> res = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();
        String temp;

        for(String s: strs) {
            char[] s_c = s.toCharArray();
            Arrays.sort(s_c);
            temp = new String(s_c);
            if(!map.containsKey(temp)) {
                map.put(temp, res.size());
                res.add(new ArrayList<>());
            }
            res.get(map.get(temp)).add(s);
        }
        return res;
    }
}

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagramMap = dict()
        for word in strs:
            sortedWord = "".join(sorted(word))
            if sortedWord not in anagramMap:
                anagramMap[sortedWord] = []
            anagramMap[sortedWord].append(word)
        
        result = []
        for sortedWord in anagramMap:
            result.append(anagramMap[sortedWord])
        return result

 
#  """
#         ["eat","tea","tan","ate","nat","bat"]

#         eat -> aet
#         tea -> aet
#         tan -> ant

#         hashmap key sortedKey -> []
#         step1: iter to list
#            each element -> sort element
#             if sort element exists in hashmap
#                append element to list
#             else
#               add new key and add the elemnt
#         m
#         o(mlogm x n) o(m x n)
# """